File: agent-skill/stackprism-site-experience/scripts/stackprism_bridge_lib/open_browser.py
import json
import os
import platform
import shutil
import subprocess
from urllib.parse import urlparse


DEFAULT_OPEN_TIMEOUT_SECONDS = 5
MAX_OPEN_TIMEOUT_SECONDS = 30
MAX_LAUNCH_PROBE_SECONDS = 1
DEFAULT_WINDOWS_PATHEXT = ".COM;.EXE;.BAT;.CMD"


def contains_nul(value):
    return isinstance(value, str) and "\0" in value or isinstance(value, list) and any(contains_nul(item) for item in value)


def has_path_separator(value):
    return "/" in value or "\\" in value


def windows_command_candidates(command, env=os.environ):
    extensions = [item for item in env.get("PATHEXT", DEFAULT_WINDOWS_PATHEXT).split(";") if item]
    if any(command.lower().endswith(extension.lower()) for extension in extensions):
        return [command]
    return [f"{command}{extension}" for extension in extensions]


def parse_open_timeout_seconds(env):
    value = env.get("STACKPRISM_BROWSER_OPEN_TIMEOUT_MS")
    if value is None or value == "":
        return True, DEFAULT_OPEN_TIMEOUT_SECONDS, None
    if not str(value).isdecimal():
        return False, None, {"reason": "invalid_open_timeout"}
    timeout_ms = int(value)
    if timeout_ms < 100 or timeout_ms > MAX_OPEN_TIMEOUT_SECONDS * 1000:
        return False, None, {"reason": "invalid_open_timeout"}
    return True, timeout_ms / 1000, None


def parse_open_config(env=os.environ):
    if any("\0" in str(env.get(key, "")) for key in ("STACKPRISM_BROWSER_OPEN_COMMAND", "STACKPRISM_BROWSER_OPEN_ARGS_JSON")):
        return False, "BRIDGE_INVALID_ENV", "Browser open environment contains NUL."
    if env.get("STACKPRISM_BROWSER_OPEN_COMMAND") and env.get("STACKPRISM_BROWSER_OPEN_ARGS_JSON"):
        try:
            if contains_nul(json.loads(env["STACKPRISM_BROWSER_OPEN_ARGS_JSON"])):
                return False, "BRIDGE_INVALID_ENV", "Browser open environment contains NUL."
        except Exception:
            pass
    return True, None, None


def resolve_browser_open_command(env=os.environ, system=None):
    command = env.get("STACKPRISM_BROWSER_OPEN_COMMAND")
    args = []
    if command:
        if env.get("STACKPRISM_BROWSER_OPEN_ARGS_JSON"):
            try:
                args = json.loads(env["STACKPRISM_BROWSER_OPEN_ARGS_JSON"])
            except Exception:
                return False, {"reason": "invalid_open_args"}
            if not isinstance(args, list) or any(not isinstance(arg, str) for arg in args):
                return False, {"reason": "invalid_open_args"}
    elif (system or platform.system()) == "Darwin":
        command = "open"
    elif (system or platform.system()) == "Windows":
        command = "rundll32.exe"
        args = ["url.dll,FileProtocolHandler"]
    else:
        command = "xdg-open"
    return True, {"command": command, "args": args}


def command_exists(command):
    if has_path_separator(command):
        if os.name == "nt":
            return any(os.path.isfile(candidate) for candidate in windows_command_candidates(command))
        return os.path.isfile(command) and os.access(command, os.X_OK)
    return shutil.which(command) is not None


def detached_popen_kwargs():
    kwargs = {"stdin": subprocess.DEVNULL, "stdout": subprocess.DEVNULL, "stderr": subprocess.DEVNULL}
    if os.name == "nt":
        kwargs["creationflags"] = getattr(subprocess, "DETACHED_PROCESS", 0) | getattr(subprocess, "CREATE_NEW_PROCESS_GROUP", 0)
    else:
        kwargs["start_new_session"] = True
    return kwargs


def open_browser(url, env=os.environ):
    ok, code, message = parse_open_config(env)
    if not ok:
        return False, {"reason": code, "message": message}
    if any(char in url for char in ("\0", "\n", "\r")):
        return False, {"reason": "invalid_url"}
    parsed_url = urlparse(url)
    if parsed_url.scheme.lower() not in ("http", "https"):
        return False, {"reason": "invalid_scheme", "allowed": ["http", "https"]}
    if parsed_url.username or parsed_url.password:
        return False, {"reason": "invalid_url"}
    if env.get("STACKPRISM_BRIDGE_NO_OPEN") == "1":
        return True, {"skipped": True}

    resolved_ok, resolved = resolve_browser_open_command(env)
    if not resolved_ok:
        return False, resolved
    timeout_ok, timeout_seconds, timeout_details = parse_open_timeout_seconds(env)
    if not timeout_ok:
        return False, timeout_details
    command = resolved["command"]
    args = resolved["args"]
    if not command_exists(command):
        return False, {"reason": "command_not_found"}

    try:
        process = subprocess.Popen([command, *args, url], **detached_popen_kwargs())
        try:
            code = process.wait(timeout=min(timeout_seconds, MAX_LAUNCH_PROBE_SECONDS))
        except subprocess.TimeoutExpired:
            return True, {}
        if code != 0:
            return False, {"reason": "open_failed", "exitCode": code}
    except FileNotFoundError:
        return False, {"reason": "command_not_found"}
    except PermissionError:
        return False, {"reason": "permission_denied"}
    except Exception:
        return False, {"reason": "spawn_failed"}
    return True, {}

File: agent-skill/stackprism-site-experience/scripts/bridge/open-browser.mjs
import { spawn } from 'node:child_process'
import { accessSync, constants, statSync } from 'node:fs'
import { delimiter, join } from 'node:path'

const DEFAULT_OPEN_TIMEOUT_MS = 5000
const MAX_OPEN_TIMEOUT_MS = 30000
const MAX_LAUNCH_PROBE_MS = 1000
const containsNul = value => (typeof value === 'string' ? value.includes('\0') : Array.isArray(value) && value.some(containsNul))
const hasPathSeparator = value => value.includes('/') || value.includes('\\')

const commandCandidates = (command, env, platform) => {
  const extensions = String(env.PATHEXT ?? process.env.PATHEXT ?? '.COM;.EXE;.BAT;.CMD')
    .split(';')
    .filter(Boolean)
  const lowerCommand = command.toLowerCase()
  if (hasPathSeparator(command)) {
    if (platform !== 'win32' || extensions.some(extension => lowerCommand.endsWith(extension.toLowerCase()))) return [command]
    return extensions.map(extension => `${command}${extension}`)
  }
  const pathEntries = String(env.PATH ?? process.env.PATH ?? '').split(delimiter).filter(Boolean)
  if (platform !== 'win32') return pathEntries.map(entry => join(entry, command))
  return pathEntries.flatMap(entry =>
    extensions.map(extension => {
      const suffix = extension.toLowerCase()
      return join(entry, lowerCommand.endsWith(suffix) ? command : `${command}${extension}`)
    })
  )
}

const commandExists = (command, env, platform) => {
  const mode = platform === 'win32' ? constants.F_OK : constants.X_OK
  for (const candidate of commandCandidates(command, env, platform)) {
    try {
      accessSync(candidate, mode)
      if (!statSync(candidate).isFile()) continue
      return true
    } catch {}
  }
  return false
}

const validateOpenUrl = url => {
  const value = String(url)
  if (value.includes('\0') || value.includes('\n') || value.includes('\r')) return { ok: false, details: { reason: 'invalid_url' } }
  let parsed
  try {
    parsed = new URL(value)
  } catch {
    return { ok: false, details: { reason: 'invalid_url' } }
  }
  if (!['http:', 'https:'].includes(parsed.protocol)) {
    return { ok: false, details: { reason: 'invalid_scheme', allowed: ['http', 'https'] } }
  }
  if (parsed.username || parsed.password) return { ok: false, details: { reason: 'invalid_url' } }
  return { ok: true }
}

export const parseOpenTimeoutMs = env => {
  const value = env.STACKPRISM_BROWSER_OPEN_TIMEOUT_MS
  if (value == null || value === '') return { ok: true, timeoutMs: DEFAULT_OPEN_TIMEOUT_MS }
  const parsed = Number(value)
  if (Number.isInteger(parsed) && parsed >= 100 && parsed <= MAX_OPEN_TIMEOUT_MS) return { ok: true, timeoutMs: parsed }
  return { ok: false, details: { reason: 'invalid_open_timeout' } }
}

export const parseOpenConfig = env => {
  for (const key of ['STACKPRISM_BROWSER_OPEN_COMMAND', 'STACKPRISM_BROWSER_OPEN_ARGS_JSON']) {
    if (String(env[key] || '').includes('\0')) {
      return { ok: false, code: 'BRIDGE_INVALID_ENV', message: 'Browser open environment contains NUL.' }
    }
  }
  if (env.STACKPRISM_BROWSER_OPEN_COMMAND && env.STACKPRISM_BROWSER_OPEN_ARGS_JSON) {
    try {
      if (containsNul(JSON.parse(env.STACKPRISM_BROWSER_OPEN_ARGS_JSON))) {
        return { ok: false, code: 'BRIDGE_INVALID_ENV', message: 'Browser open environment contains NUL.' }
      }
    } catch {}
  }
  return { ok: true }
}

export const resolveBrowserOpenCommand = (env = process.env, platform = process.platform) => {
  let command = env.STACKPRISM_BROWSER_OPEN_COMMAND
  let args = []
  if (command) {
    if (env.STACKPRISM_BROWSER_OPEN_ARGS_JSON) {
      try {
        args = JSON.parse(env.STACKPRISM_BROWSER_OPEN_ARGS_JSON)
      } catch {
        return { ok: false, details: { reason: 'invalid_open_args' } }
      }
      if (!Array.isArray(args) || args.some(arg => typeof arg !== 'string')) {
        return { ok: false, details: { reason: 'invalid_open_args' } }
      }
    }
  } else if (platform === 'darwin') {
    command = 'open'
  } else if (platform === 'win32') {
    command = 'rundll32.exe'
    args = ['url.dll,FileProtocolHandler']
  } else {
    command = 'xdg-open'
  }
  return { ok: true, command, args }
}

const waitForLaunchProbe = (child, timeoutMs) =>
  new Promise(resolve => {
    let settled = false
    const finish = result => {
      if (settled) return
      settled = true
      clearTimeout(timer)
      child.off('error', onError)
      child.off('exit', onExit)
      resolve(result)
    }
    const onError = error => {
      finish({ ok: false, details: { reason: error?.code === 'ENOENT' ? 'command_not_found' : 'spawn_failed' } })
    }
    const onExit = code => {
      finish(code === 0 ? { ok: true } : { ok: false, details: { reason: 'open_failed', exitCode: code } })
    }
    const timer = setTimeout(() => finish({ ok: true }), Math.min(timeoutMs, MAX_LAUNCH_PROBE_MS))
    timer.unref?.()
    child.once('error', onError)
    child.once('exit', onExit)
  })

export const openBrowser = async (url, env = process.env, platform = process.platform) => {
  const openConfig = parseOpenConfig(env)
  if (!openConfig.ok) return { ok: false, details: { reason: openConfig.code, message: openConfig.message } }
  const openUrl = validateOpenUrl(url)
  if (!openUrl.ok) return openUrl

  if (env.STACKPRISM_BRIDGE_NO_OPEN === '1') return { ok: true, skipped: true }

  const resolved = resolveBrowserOpenCommand(env, platform)
  if (!resolved.ok) return resolved
  const timeout = parseOpenTimeoutMs(env)
  if (!timeout.ok) return { ok: false, details: timeout.details }
  const { command, args } = resolved
  if (!commandExists(command, env, platform)) return { ok: false, details: { reason: 'command_not_found' } }

  try {
    const child = spawn(command, [...args, url], { detached: true, stdio: 'ignore', shell: false })
    const launched = await waitForLaunchProbe(child, timeout.timeoutMs)
    child.unref()
    return launched
  } catch {
    return { ok: false, details: { reason: 'spawn_failed' } }
  }
}
